File: year2/mpp/java/MotorcycleGrpcClient/MotorcycleServer/src/main/java/services/MotorcycleServiceImpl.java
package services;

import dtos.ParticipantDto;
import dtos.RaceDto;
import domain.Participant;
import domain.Race;
import domain.User;
import repositories.interfaces.ParticipantRepository;
import repositories.interfaces.RaceRepository;
import repositories.interfaces.TeamRepository;
import repositories.interfaces.UserRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MotorcycleServiceImpl implements MotorcycleService {
    private UserRepository userRepository;
    private RaceRepository raceRepository;
    private TeamRepository teamRepository;
    private ParticipantRepository participantRepository;

    private List<MotorcycleObserver> observers = new ArrayList<>();

    public MotorcycleServiceImpl(UserRepository userRepository, RaceRepository raceRepository, TeamRepository teamRepository, ParticipantRepository participantRepository) {
        this.userRepository = userRepository;
        this.raceRepository = raceRepository;
        this.teamRepository = teamRepository;
        this.participantRepository = participantRepository;
    }

    public MotorcycleServiceImpl() {
    }

    public void setUserRepository(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public void setRaceRepository(RaceRepository raceRepository) {
        this.raceRepository = raceRepository;
    }

    public void setTeamRepository(TeamRepository teamRepository) {
        this.teamRepository = teamRepository;
    }

    public void setParticipantRepository(ParticipantRepository participantRepository) {
        this.participantRepository = participantRepository;
    }

    public synchronized Optional<User> authenticate(String username, String password) {
        return userRepository.findByUsernameAndPassword(username, password);
    }

    public synchronized RaceDto getRaceByCapacity(Integer capacity) {
        Integer raceId = raceRepository.findIdByCapacity(capacity);
        Optional<Race> race = raceRepository.findOne(raceId);
        List<Participant> participants = participantRepository.findByRace(raceId);
        return new RaceDto(race.get(), participants.size());
    }

    public synchronized List<RaceDto> getRaceInfos() {
        List<RaceDto> races = new ArrayList<>();
        for (Race race : raceRepository.findAll()) {
            List<Participant> participants = participantRepository.findByRace(race.getId());
            races.add(new RaceDto(race, participants.size()));
        }
        return races;
    }

    public synchronized Iterable<Race> getRaces() {
        return raceRepository.findAll();
    }

    public synchronized List<ParticipantDto> getRaceParticipantsByTeam(String teamName) {
        List<ParticipantDto> participants = new ArrayList<>();
        Integer teamId = teamRepository.findIdByName(teamName);
        if (teamId == 0) {
            return participants;
        }
        for (Participant participant : participantRepository.findByTeam(teamId)) {
            Optional<Race> race = raceRepository.findOne(participant.getRaceId());
            race.ifPresent(value -> participants.add(new ParticipantDto(value, participant)));
        }
        return participants;
    }

    public synchronized void registerParticipant(String name, String teamName, Integer capacity) {
        Integer teamId = teamRepository.findIdByName(teamName);
        Integer raceId = raceRepository.findIdByCapacity(capacity);
        Participant participant = new Participant(name, raceId, teamId);
        participantRepository.save(participant);

        ExecutorService executor = Executors.newFixedThreadPool(5);
        executor.execute(() -> {
            observers.forEach(obs -> {
                try {
                    obs.registeredParticipant(participant);
                } catch (Exception e) {
                    System.out.println(e.getMessage());
                }
            });
        });
        executor.shutdown();
    }

    @Override
    public void logOut(User user) {
        // notify friends
    }

    @Override
    public void addMotorcycleObserver(MotorcycleObserver observer) {
        this.observers.add(observer);
    }

    @Override
    public void removeMotorcycleObserver(MotorcycleObserver observer) {
        this.observers.remove(observer);
    }
}

File: year2/mpp/c#/MotorcycleGrpcServer/MotorcycleServer/Services/MotorcycleServiceImpl.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using MotorcycleCommon.Services;
using MotorcycleContest.Domain.Dtos;
using MotorcycleContest.Domain.Entities;
using MotorcycleContest.Repositories.Interfaces;

namespace MotorcycleContest.Services
{
    public class MotorcycleServiceImpl : IMotorcycleService
    {
        private readonly IUserRepository _userRepository;
        private readonly IRaceRepository _raceRepository;
        private readonly ITeamRepository _teamRepository;
        private readonly IParticipantRepository _participantRepository;

        private List<IMotorcycleObserver> _observers = new List<IMotorcycleObserver>();

        public MotorcycleServiceImpl(IUserRepository userRepository, IRaceRepository raceRepository,
            ITeamRepository teamRepository, IParticipantRepository participantRepository)
        {
            this._userRepository = userRepository;
            this._raceRepository = raceRepository;
            this._teamRepository = teamRepository;
            this._participantRepository = participantRepository;
        }

        public User Authenticate(string username, string password)
        {
            return _userRepository.FindByUsernameAndPassword(username, password);
        }

        public RaceDto GetRaceByCapacity(int capacity)
        {
            int raceId = _raceRepository.FindIdByCapacity(capacity);
            Race race = _raceRepository.FindOne(raceId);
            List<Participant> participants = _participantRepository.FindByRace(race.ID);
            return new RaceDto(race, participants.Count);
        }

        public IEnumerable<Race> GetRaces() {
            return _raceRepository.FindAll();
        }

        public IEnumerable<RaceDto> GetRaceInfos()
        {
            List<RaceDto> races = new List<RaceDto>();
            foreach (var race in _raceRepository.FindAll())
            {
                List<Participant> participants = _participantRepository.FindByRace(race.ID);
                races.Add(new RaceDto(race, participants.Count));
            }

            return races;
        }

        public List<ParticipantDto> GetRaceParticipantsByTeam(string teamName) {
            List<ParticipantDto> participants = new List<ParticipantDto>();
            int teamId = _teamRepository.FindIdByName(teamName);
            if (teamId == 0) {
                return participants;
            }
            foreach (Participant participant in _participantRepository.FindByTeam(teamId)) {
                Race race = _raceRepository.FindOne(participant.RaceId);
                if (race != null)
                {
                    participants.Add(new ParticipantDto(race, participant));
                }
            }
            return participants;
        }

        public void RegisterParticipant(string name, string teamName, int capacity) {
            int teamId = _teamRepository.FindIdByName(teamName);
            int raceId = _raceRepository.FindIdByCapacity(capacity);
            Participant participant = _participantRepository.Save(new MotorcycleContest.Domain.Entities.Participant(name, raceId, teamId));

            Task.Run(() =>
            {
                _observers.ForEach(o =>
                {
                    o.RegisteredParticipant(participant);
                });
            });
        }

        public void logOut(User user)
        {
            
        }

        public void AddMotorcycleObserver(IMotorcycleObserver observer)
        {
            _observers.Add(observer);
        }

        public void RemoveMotorcycleObserver(IMotorcycleObserver observer)
        {
            _observers.Remove(observer);
        }
    }
}